source/core_mqtt_serializer.c

```diff
@@ -2677,4 +4121,382 @@ MQTTStatus_t MQTT_ProcessIncomingPacketTypeAndLength( const uint8_t * pBuffer,
+            }
+        }
+
+        if( status == MQTTSuccess )
+        {
+            /* Add the lengths of the will message and topic name if provided. */
+            if( pWillInfo != NULL )
+            {
+                connectPacketSize += pWillInfo->propertyLength;
+                connectPacketSize += remainingLengthEncodedSize( pWillInfo->propertyLength );
+                connectPacketSize += pWillInfo->topicNameLength + sizeof( uint16_t ) +
+                                     pWillInfo->payloadLength + sizeof( uint16_t );
+            }
+
+            /* Add the lengths of the user name and password if provided. */
+            if( pConnectInfo->pUserName != NULL )
+            {
+                connectPacketSize += pConnectInfo->userNameLength + sizeof( uint16_t );
+            }
+
+            if( pConnectInfo->pPassword != NULL )
+            {
+                connectPacketSize += pConnectInfo->passwordLength + sizeof( uint16_t );
+            }
+
+            /* At this point, the "Remaining Length" field of the MQTT CONNECT packet has
+             * been calculated. */
+            remainingLength = connectPacketSize;
+
+            /* Calculate the full size of the MQTT CONNECT packet by adding the size of
+             * the "Remaining Length" field plus 1 byte for the "Packet Type" field. */
+            connectPacketSize += 1U + remainingLengthEncodedSize( connectPacketSize );
+
+            /**
+             * 268,435,455 - max remaining length  according to spec MQTT-v5;
+             *
+             * */
+            if( remainingLength > MQTT_MAX_REMAINING_LENGTH )
+            {
+                status = MQTTBadParameter;
+            }
+        }
+
+        if( status == MQTTSuccess )
+        {
+            *pRemainingLength = remainingLength;
+            *pPacketSize = connectPacketSize;
+
+            LogDebug( ( "CONNECT packet remaining length=%lu and packet size=%lu.",
+                        ( unsigned long ) *pRemainingLength,
+                        ( unsigned long ) *pPacketSize ) );
+        }
+
+        return status;
+    }
+
+    uint8_t * MQTT_SerializePublishProperties( const MQTTPublishInfo_t * pPublishInfo,
+                                               uint8_t * pIndex )
+    {
+        uint8_t * pIndexLocal = pIndex;
+
+        pIndexLocal = encodeRemainingLength( pIndexLocal, pPublishInfo->propertyLength );
+
+        /*Serialize the will delay if provided.*/
+
+        if( pPublishInfo->willDelay != 0U )
+        {
+            *pIndexLocal = MQTT_WILL_DELAY_ID;
+            pIndexLocal++;
+            pIndexLocal[ 0 ] = UINT32_BYTE3( pPublishInfo->willDelay );
+            pIndexLocal[ 1 ] = UINT32_BYTE2( pPublishInfo->willDelay );
+            pIndexLocal[ 2 ] = UINT32_BYTE1( pPublishInfo->willDelay );
+            pIndexLocal[ 3 ] = UINT32_BYTE0( pPublishInfo->willDelay );
+            pIndexLocal = &pIndexLocal[ 4 ];
```

Comment:
Why are we incrementing the pointer in this different way than usual? Any refernce to the existing code?